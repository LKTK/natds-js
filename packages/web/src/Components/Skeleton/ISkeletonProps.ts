import { SkeletonProps } from "@material-ui/lab/Skeleton";
import { ISizes } from "@naturacosmeticos/natds-styles";

export interface ISkeletonProps extends Pick<SkeletonProps,
  "animation" |
  "children" |
  "variant"
  > {

  /**
   * Height of the skeleton.
   *
   *Useful when you want to adapt the skeleton to a element - for instance, a card.
   */
  height?: keyof ISizes,

  /**
   * Width of the skeleton. Useful when the skeleton is inside an inline element with no width of its own.
   */
  width?: keyof ISizes
}
