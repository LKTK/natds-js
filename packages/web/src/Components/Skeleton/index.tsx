import MaterialSkeleton from "@material-ui/lab/Skeleton";

export { SkeletonProps as ISkeletonProps } from "@material-ui/lab/Skeleton";

export default MaterialSkeleton;
