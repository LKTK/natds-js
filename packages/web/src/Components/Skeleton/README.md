## Usage

```jsx
{item ? (
  <img alt={item.title} height= src={item.src} />
) : (
  <Skeleton variant={"rect"} />
)}
```

## Inferring dimensions

In addition to accepting `height` and `width` props, the component can also infer the dimensions.

It works well when it comes to typography as its height is set using `em` units.

<Preview>
  <Story name="Skeleton with Typography">
    <ComponentWithLabel componentList={forTypography} headingText={"With Typography"} />
  </Story>
</Preview>

```jsx
import React from "React";
import Typography from "@naturacosmeticos/natds-web/dist/Typography";

export const SkeletonTypography = () => {

  const loading = false;

  return (
    <div>
      <Typography variant={"h1"}>
        { loading ? <Skeleton /> : "h1" }
      </Typography>
    </div>
  );
}
```

```jsx
import React from "react";
import Avatar from "@naturacosmeticos/natds-web/dist/Avatar";
import Skeleton from "@naturacosmeticos/natds-web/dist/Skeleton";

export const SkeletonChildrenDemo = () => {

  const loading = false;

  return (
    <div>
      { loading ? (
        <Skeleton variant={"circle"}>
          <Avatar />
        </Skeleton>
      ) : (
        <Avatar src={"https://"} />
      )}
    </div>
  );

}
```

## Accessibility

Skeleton screens provide an alternative to the traditional spinner methods.

Rather than showing an abstract widget, skeleton screens create anticipation of what is to come and reduce cognitive load.

## Skeleton API

The API documentation of the Skeleton component. Learn more about the props.

The `ref` is forwarded to the root element.

Any other props supplied will be provided to the root element.
