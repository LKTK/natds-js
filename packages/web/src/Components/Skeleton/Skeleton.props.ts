import { SkeletonProps } from "@material-ui/lab/Skeleton";
import { SizeProp } from "../shared/SizeProp";

export interface ISkeletonProps extends Pick<SkeletonProps,
  "animation" |
  "children" |
  "variant"
  > {

  /**
   * Height of the skeleton.
   *
   * Useful when you want to adapt the skeleton to a element - for instance, a card.
   *
   * @optional
   * @type SizeProp
   */
  height?: SizeProp,

  /**
   * Width of the skeleton.
   *
   * Useful when the skeleton is inside an inline element with no width of its own.
   *
   * @optional
   * @type SizeProp
   */
  width?: SizeProp
}
