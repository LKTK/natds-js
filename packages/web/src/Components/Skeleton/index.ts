import Skeleton from "./Skeleton";

export { ISkeletonProps } from "./ISkeletonProps";
export { Skeleton };
export default Skeleton;
