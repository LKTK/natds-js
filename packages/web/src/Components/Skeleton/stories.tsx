import * as React from "react";
import { tokens } from "@naturacosmeticos/natds-styles";
import Skeleton from ".";
import {IStorySections} from "../types/IStorySections";
import Avatar from "../Avatar";
import Typography, {TypographyVariant} from "../Typography";

export const variants: IStorySections = [
  {
    component: <Skeleton variant={"text"} />,
    title: "Text",
  },
  {
    component: <Skeleton variant={"circle"} height={tokens.sizes.semix} width={tokens.sizes.semix} />,
    title: "Circle: semix size (40px)",
  },
  {
    component: <Skeleton variant={"rect"} height={tokens.sizes.hugex} width={tokens.sizes.veryhuge} />,
    title: "Rect",
  },
];

export const animations: IStorySections = [
  {
    component: <Skeleton width={tokens.sizes.veryhuge} />,
    title: "Default animation (pulsate)",
  },
  {
    component: <Skeleton animation={false} width={tokens.sizes.veryhuge} />,
    title: "No animation",
  },
  {
    component: <Skeleton animation="wave" width={tokens.sizes.veryhuge} />,
    title: "Wave animation",
  },
];

const typographyVariants : TypographyVariant[] = ["h1",
  "h3",
  "body1",
  "caption"];

export const forTypography: IStorySections = typographyVariants.map((variant) => ({
  component: <Typography variant={variant}><Skeleton /></Typography>,
  title: variant,
}));

const avatarAltText = "Nat Natura Avatar";
const avatarUrl = "https://pbs.twimg.com/profile_images/1285932878795333632/Wl95KE_m_400x400.jpg";

export const forAvatar : IStorySections = [
  {
    component: <Skeleton variant={"circle"}><Avatar alt={avatarAltText} size={"huge"} src={avatarUrl} /></Skeleton>,
    title: "Avatar (loading)",
  },
  {
    component: <Avatar alt={avatarAltText} size={"huge"} src={avatarUrl} />,
    title: "Avatar (loaded)",
  },
];
