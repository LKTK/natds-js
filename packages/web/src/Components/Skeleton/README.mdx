import { Meta, Preview, Props, Story } from '@storybook/addon-docs/blocks';
import { ComponentWithLabel, getComponentParams, getDefaultDecorators } from '../../../.storybook';

import Skeleton from '.'
import { animations, forAvatar, forTypography, variants } from './stories'

<Meta component={Skeleton} parameters={getComponentParams(["Skeleton"])} title="Components|Skeleton" />

# Skeleton

Display a placeholder preview of your content before the data gets loaded to reduce load-time frustration.

The data for your components might not be immediately available. You can increase the perceived performance for users
by using skeletons. It feels like things are happening immediately, then the information is incrementally displayed on
the screen. (Cf. [Avoid The Spinner](https://www.lukew.com/ff/entry.asp?1797))

The component can be used *directly in your components*. For instance:

```jsx
{item ? (
  <img alt={item.title} height= src={item.src} />
) : (
  <Skeleton variant={"rect"} />
)}
```

## Variants

The component supports 3 shape variants.

### Text

<Preview>
  <Story name="Skeleton variants">
    <ComponentWithLabel componentList={variants} headingText={"Variants"} />
  </Story>
</Preview>

```jsx
import React from "react";
import Skeleton from "@naturacosmeticos/natds-web/dist/Components"

export const Variants = () => (
  <div>
    <Skeleton variant={"text"} />
    <Skeleton variant={"circle"} height={40} width={40} />
    <Skeleton variant={"rect"} height={118} width={210} />
  </div>
)
```

## Animations

By default, the skeleton pulsate, but you can change the animation for a wave or disable it entirely.

<Preview>
  <Story name="Skeleton animations">
    <ComponentWithLabel componentList={animations} headingText={"Animations"} />
  </Story>
</Preview>

```jsx
import React from "react";
import Skeleton from "@naturacosmeticos/natds-web/dist/Components"

export const Animations = () => (
  <div>
      <Skeleton />
      <Skeleton animation={false} />
      <Skeleton animation="wave" />
  </div>
)
```

## Inferring dimensions

In addition to accepting `height` and `width` props, the component can also infer the dimensions.

It works well when it comes to typography as its height is set using `em` units.

<Preview>
  <Story name="Skeleton with Typography">
    <ComponentWithLabel componentList={forTypography} headingText={"With Typography"} />
  </Story>
</Preview>

```jsx
import React from "React";
import Typography from "@naturacosmeticos/natds-web/dist/Typography";

export const SkeletonTypography = () => {

  const loading = false;

  return (
    <div>
      <Typography variant={"h1"}>
        { loading ? <Skeleton /> : "h1" }
      </Typography>
    </div>
  );
}
```

However, when it comes to other components, you may not want to repeat the height and width.

In these instances, you can pass `children` and it will infer its width and height from them.

<Preview>
  <Story name="Skeleton with Avatar">
    <ComponentWithLabel componentList={forAvatar} header={"With Avatar"} />
  </Story>
</Preview>

```jsx
import React from "react";
import Avatar from "@naturacosmeticos/natds-web/dist/Avatar";
import Skeleton from "@naturacosmeticos/natds-web/dist/Skeleton";

export const SkeletonChildrenDemo = () => {

  const loading = false;

  return (
    <div>
      { loading ? (
        <Skeleton variant={"circle"}>
          <Avatar />
        </Skeleton>
      ) : (
        <Avatar src={"https://"} />
      )}
    </div>
  );

}
```

## Accessibility

Skeleton screens provide an alternative to the traditional spinner methods.

Rather than showing an abstract widget, skeleton screens create anticipation of what is to come and reduce cognitive load.

## Skeleton API

The API documentation of the Skeleton component. Learn more about the props.

### Import

```jsx
import Skeleton from "@naturacosmeticos/natds-web/dist/Skeleton";
// or
import { Skeleton } from "@naturacosmeticos/natds-web";
```

### Props

The `ref` is forwarded to the root element.

Any other props supplied will be provided to the root element.
