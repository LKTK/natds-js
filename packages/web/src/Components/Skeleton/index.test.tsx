import * as React from "react";
import * as renderer from "react-test-renderer";
import Skeleton from ".";

describe("Skeleton component", () => {
  it("renders correctly", () => {
    const actual: renderer.ReactTestRenderer = renderer.create(<Skeleton />);

    expect(actual).toMatchSnapshot();
  });
});
