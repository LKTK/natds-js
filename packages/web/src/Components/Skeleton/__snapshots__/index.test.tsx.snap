// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Skeleton component renders correctly 1`] = `
<span
  className="MuiSkeleton-root MuiSkeleton-text MuiSkeleton-pulse"
  style={
    Object {
      "height": undefined,
      "width": undefined,
    }
  }
/>
`;
